---
title: "TCGAbrowser workflow"
author: "Phil Cheng"
date: "`r doc_date()`"
package: "`r pkg_ver('TCGAbrowser')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{TCGAbrowser workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction
The Cancer Genome Atlas has provided a comprehensive omics study of 33 cancers. The National Cancer Institute (NCI) has made this data pubically available on the Genomic Data Commons [(GDC)](https://gdc-portal.nci.nih.gov/) Data Portal. There are several R packages that download the data, `r Biocpkg("RTCGAToolbox")`, `r Biocpkg("RTCGA")`, `r Biocpkg("TCGAbiolinks")`, `r CRANpkg("TCGA2STAT")`, `r Biocpkg("MultiAssayExperiment")`, and [TCGA-Assembler](http://www.compgenome.org/TCGA-Assembler/).

The motivation for TCGAbrowser was to provide convenience functions and wrappers to analyze the omics datasets by gene expression or mutation subsetting. Typically, biologists are studying one gene in one cancer by overexpression and/or knockdown experiments. TCGAbrowser was designed to subset the datasets by gene expression or gene mutation and perform differential gene expresion, mutation, copy number, protein and survival analysis. 

# Installation
To install TCGAbrowser use the code below.
```{r, message = FALSE, warning = FALSE}
require(devtools)
#devtools::install_github("pcheng84/TCGAbrowser", ref = "multiAssayExperiment")
require(TCGAbrowser)
```

The workflow for TCGAbrowser is shown in the graphic below.

```{r, out.width = "50%", echo = FALSE}
knitr::include_graphics("workflow2.png")
```

TCGAbrowser functions were designed either to subset the data, perform statistical analysis or for plotting. 

Subset functions include `rnasubset` and `mutsubset` which that subset the data based on RNA expression and mutation status respectively. 

Analysis functions include: `diffmut` for differential mutation analysis; `diffcp` for differential copy number analysis; `diffrna` for differential RNA expression analysis; `rnagsva` and `rnagsvasig` for differential GSVA analysis; `rnareact` for reactome pathway analysis on the differential RNA expression analysis; `diffrppa` for differential RPPA analysis, `diffmethyl` for differential methylation analysis.

Plotting functions include: `plotsurv` for Kaplan-Meier survival plots; `plotmut` for an Oncoprint-like plot of mutation data; `plotcp` for an Oncoprint-like plot of copy number data; `plotrnaheat` for a heatmap of differential expressed genes; `plotgsvaheat` for a heatmap of differential expressed pathways; `plotreact` for Reactome dotplots, enrichment plots and cnetmap plots; `plotrppaheat` for a heatmap of differential expressed proteins.


#Data download and formating
This vignette assumes the user will use one of the R packages mentioned above to download the TCGA data. In this example, we used `r Biocpkg("curatedTCGAData")` to download the data. We will transform the mutation data into a binary matrix with `qreduceTCGA` from `r Biocpkg("TCGAutils")` and subset the data for just tumor samples using `splitAssays`. 

```{r, message = FALSE, warning = FALSE}
library(curatedTCGAData)
library(TCGAutils)

lusc <- curatedTCGAData("LUSC", c("Mutation", "RNASeq2GeneNorm", "GISTIC_AllbyGene", "RPPAArray", "Methylation_methyl450"), FALSE)

#convert genome to hg19 format
genome(lusc[[2]]) <- vapply(genome(lusc[[2]]), translateBuild, character(1L))
seqlevelsStyle(lusc[[2]]) <- "NCBI"

#convert mutation into binary matrix
lusc2 <- qreduceTCGA(lusc)

#remove normal samples and just focus on primaries
tnmae <- splitAssays(lusc2, c("01", "11"))

lusc_t <- tnmae[, , grep("^01", names(tnmae))]

```

#Subset by gene expression workflow
`rnasubset` needs 3 parameters: a multiassayexperiment object, a gene and a percentage cutoff. This function will return a multiassayexperiment object with an assay called "Cohort". The "Cohort" assay contains all samples from the RNAseq dataset and has 3 rows, the first row identifies if the sample belongs in the "low", "middle", or "high" gene expression category, the second row is the rank of gene across all samples, and the third row are the gene expression values themselves. 

In this example, we will compare patients in the top 10% of EGFR expression and to the bottom 10% of EGFR expression.

```{r}
lusc_egfr <- rnasubset(lusc_t, "EGFR", 10)
```

##Expression plotting
The output table from `rnasubset` can be used for downstream analysis and plotting functions. `plotgenelevel` will create a ggplot style dotplot  These two functions will plot the gene expression of your gene of interest across all samples with coloring for the low, middle and high groups. 

```{r}
plotgenelevel(lusc_egfr)

```

##Survival Analysis
`plotsurv` needs 2 parameters: the output multiassayexperiment object from `rnasubset` and the gene name. `plotsurv` is a wrapper for `ggsurvplot` from `r CRANpkg("survminer")`. It will plot the Kaplan-Meier curves for the high gene and low gene expression groups and report the p value from the log-rank test.   

```{r}
genesurv(lusc_egfr, "EGFR")
```

##Differential gene expression analysis
`diffrna` needs 2 parameters: the output multiassayexperiment object from `rnasubset`. `diffrna` is a wrapper for `voom` from `r Biocpkg("limma")`. It will calculate the genes differentially expressed between the high gene and low gene expression groups.

```{r}
egfr_deg <- diffrna(lusc_egfr)
head(egfr_deg)
```

`plotrnaheat` needs 4 parameters: the output multiassayexperiment object from `rnasubset`, the output data.frame from `diffrna`, the gene name, and the number of genes to plot for the heatmap. `plotrnaheat` is a wrapper for `Heatmap` from`r CRANpkg("ComplexHeatmap")`. It will plot a heatmap of the top 100 most significant differentially expressed genes, draw a colored column bar to indicate the high gene and low gene expression groups, and a colored column bar to indicate the rank of the gene of interest.  

```{r}
plotrnaheat(lusc_egfr, egfr_deg, "EGFR", 100)
```

##Reactome pathway analysis
`rnareact` needs 1 parameter: the output table from `diffrna`. `rnareact` is wrapper for `enrichPathway` from `r Biocpkg("ReactomePA")`. It will calculate enriched pathways from Reactome database. 

`plotreact` needs 4 parameters: the output table from `rnareact`, the output table from `diffrna`, the graph selection which can be 'dot', 'map', or 'cnet', and the number of pathways for plotting. `plotreact` is a wrapper for `dotplot`, `enrichMap`, and `cnetplot` from `r Biocpkg("ReactomePA")`. 

```{r}
egfr_react <- rnareact(egfr_deg)

#dotplot
plotreact(egfr_react, egfr_deg, "dot", 15)

#pathway plot
plotreact(egfr_react, egfr_deg, "map", 15)

#cnet plot
plotreact(egfr_react, egfr_deg, "cnet", 5)
```

##Differential mutation
`diffmut` needs a MultiAssayExperiment object from `rnasubset` or `mutsubset`. The MultiAssayExperiment should contain an assay with a simplified Mutation from `qreduceTCGA` or a binary matrix representing mutated and wild-type genes. It will perform a chi-squared test for the proportions of mutated and wild-type genes between the high gene and low gene expression groups from `rnasubset` or mutated and wild-type groups from `mutsubset`. Since we use a binary matrix as a representation of mutation, we assume is any mutation in a gene would have the same effect. 

`plotmut` needs 4 parameters: the MultiAssayExperiment object from `rnasubset` or `mutsubset`, the output table from `diffmut`, the gene name, and the number of genes to plot. It will create an Oncoprint-like plot of the differentially mutated genes and your gene of interest.

```{r}
egfr_dm <- diffmut(lusc_egfr)
plotmut(lusc_egfr, egfr_dm, "EGFR", 20)
```

##Differential copy number analysis
`diffcp` needs a MultiAssayExperiment object from `rnasubset` or `mutsubset`. The MultiAssayExperiment should contain a GISTIC assay where the copy number values are represented by -2, -1, 0, 1, and 2. It will perform a chi-squared test between copy number gain and copy number neutral and another chi-squared test for copy number neutral and copy number loss. 

`plotcp` needs 4 parameters: the MultiAssayExperiment object from `rnasuset` or `mutsubset`, the output table from `diffcp`, the gene name, and the number of genes to plot. It will create an Oncoprint-like plot of the copy number differences and your gene of interest.

```{r}
egfr_dcp <- diffcp(lusc_egfr)
plotcp(lusc_egfr, egfr_dcp, "EGFR", 20)
```

##Differential methylation analyis
Methylation data from TCGA usually comes from the Illumina Methyl450k arrays. The data is quite large and to get an overview of the differentially methylated regions, we collapse the methylation data to island level using the annotation from `r Biocpkg(IlluminaHumanMethylation450kanno.ilmn12.hg19)`. `methylprep` will take the median of all the probes in the island and return MultiAssayExperiment object with an assay named "SimpleMethyl" which contains a matrix of the median B values for each island per sample. The rownames of "SimpleMethyl" are the genomic coordinates of each island. This new MAE object can then be passed to `diffmethyl` where it will do an arcsin transformation of the B values and uses `lmfit` from `r Biocpkg(limma)` to get the differential methylated islands. 

```{r}
lusc_t <- methylprep(lusc_t)
#subset samples by top and bottom 10% of EGFR expression
lusc_t.egfr <- rnasubset(lusc_t, "EGFR", 10)
#look for differentially methylated islands
egfr_meth_deg <- diffmeth(lusc_t.egfr)
```

##Subsetting by mutation
`mutsubset` needs 2 parameters: a MultiAssayExperiment object and a gene name. The MultiAssayExperiment should contain an assay with a simplified Mutation from `qreduceTCGA` or a binary matrix representing mutated and wild-type genes. It will return a new MultiAssayExperiment object with an assay name "Cohort" . The "Cohort" assay contains all samples from the Mutation dataset and has 1 row identifying if the sample has the mutation in your gene of interest.

```{r}
lusc_t.egfr <- mutsubset(lusc_t, "EGFR")
```




#Creating Multiassayexperiment from TCGAbiolinks
```{r eval = FALSE, echo = TRUE}
library(data.table)

library(TCGAbiolinks)
setwd("../")
query <- GDCquery(project = "TCGA-SKCM",
                  data.category = "Clinical",
                  file.type = "xml")
GDCdownload(query)
skcm.clin <- GDCprepare_clinic(query, clinical.info = "patient")
skcm.clin.f <- GDCprepare_clinic(query, clinical.info = "follow_up")
setDT(skcm.clin.f)
setDT(skcm.clin)
skcm.clin[, days := ifelse(is.na(days_to_death), days_to_last_followup, days_to_death)]
skcm.clin[, TCGA_days := days - days_to_submitted_specimen_dx]

query1 <- GDCquery(project = "TCGA-SKCM",
                  data.category = "Copy Number Variation",
                  data.type = "Gene Level Copy Number Scores",
                  access="open")
GDCdownload(query1)
skcm_gis <- GDCprepare(query1)

query2 <- GDCquery(project = "TCGA-SKCM",
                  data.category = "Copy Number Variation",
                  data.type = "Copy Number Segment",
                  access="open")
GDCdownload(query2, files.per.chunk = 20)
skcm_gis2 <- GDCprepare(query2)
setDT(skcm_gis)
skcm_maf <- GDCquery_Maf("SKCM", pipelines = "mutect2")
setDT(skcm_maf)
skcm_rna <- GDCquery(project = "TCGA-SKCM",
                     data.category = "Transcriptome Profiling",
                     data.type = "Gene Expression Quantification",
                     workflow.type = "HTSeq - Counts")

GDCdownload(skcm_rna, method = "api", files.per.chunk = 20)
skcm_rna2 <- GDCprepare(skcm_rna)
```
