#' diffmut function
#'
#' Computes differentially mutated genes between two subgroups selected by `rnasubset` or `mutsubset` by Chi-squared test
#'
#' @param mae MultiAssayExperiment object. Must contain assays simplified Mutation (generated by `qreduceTCGA`)
#' or a binary matrix of mutated genes by sample, assay should contain "Mutation" in the name,
#' and Cohort (from `rnasubset` or `mutsubset`)
#'
#' @importFrom MultiAssayExperiment intersectColumns
#' @import data.table
#' @importFrom reshape2 dcast
#' @return data frame with frequency of mutation in high and low group with p-value from Chi-squared test
#'
#' @examples
#' library(curatedTCGAData)
#' library(TCGAutils)
#' lusc <- curatedTCGAData("LUSC", c("Mutation", "RNASeq2GeneNorm"), FALSE)
#' mut_assay <- grep("Mutation", names(lusc))
#' genome(lusc[[mut_assay]]) <- vapply(genome(lusc[[mut_assay]]), translateBuild, character(1L))
#' seqlevelsStyle(lusc[[mut_assay]]) <- "NCBI"
#' lusc2 <- qreduceTCGA(lusc)
#'
#' #remove normal samples and just focus on primaries
#' tnmae <- splitAssays(lusc2, c("01", "11"))
#'
#' lusc.t <- tnmae[, , grep("^01", names(tnmae))]
#' #split tumor and normal samples
#' lusc_tn <- splitAssays(lusc2, c("01", "11"))
#' #remake MultiAssayExperiment with only primary tumor samples
#' lusc_t <- lusc_tn[, , grep("^01", names(lusc_tn))]
#' lusc_t.egfr <- rnasubset(lusc_t, "EGFR", 10)
#' egfr_dm <- diffmut(lusc_t.egfr)
#'
#' @export
#'

diffmut <- function(mae) {
  #check only one mutation and one cohort assay in the multiassayexperiment object
  stopifnot(class(mae) == "MultiAssayExperiment", any(grepl("Mutation", names(mae))), any(grepl("Cohort", names(mae))))
  if(length(grep("Mutation", names(mae))) > 1)
    stop("Only one mutation assay is allowed")
  if(length(grep("Cohort", names(mae))) > 1)
    stop("Only one cohort comparison is allowed")

  # Find which assays contain mutation and expression level
  mut_assay <- grep("Mutation", names(mae))
  exp_assay <- grep("Cohort", names(mae))

  mae2 <- intersectColumns(mae[, , c(mut_assay, exp_assay)])

  annot <- dcast(as.data.frame(sampleMap(mae2)), primary ~ assay, value.var = "colname")
  lvl2 <- data.frame(Cohort = colnames(mae2[[2]]), Level = mae2[[2]][1,])
  lvl3 <- merge(annot, lvl2, by = "Cohort")
  lvl.high <- lvl3[lvl3$Level == "high", grep("Mutation", colnames(lvl3))]
  lvl.low <- lvl3[lvl3$Level == "low", grep("Mutation", colnames(lvl3))]

  #only proceed if both high and low have more than 1 sample
  if(length(lvl.high) == 0 | length(lvl.low) == 0)
    return("Not enough mutation profiles") else {
  mut <- assay(mae2[[1]])

  #if only one sample has mutation data just assign it
  if(length(lvl.high)  == 1)
    N.high <- mut[, lvl.high] else
      N.high <- rowSums(mut[, lvl.high])
  if(length(lvl.low) == 1)
    N.low <- mut[, lvl.low] else
      N.low <- rowSums(mut[, lvl.low])

  #calculates most mutated genes in the overlap patients
  m1 <- data.table(Gene = rownames(mut),
                   N.high = N.high,
                   N.hightotal = length(lvl.high),
                   N.low = N.low,
                   N.lowtotal = length(lvl.low))

  m2 <- m1[N.high + N.low != 0 & N.high + N.low != 1]

  #return empty table if no differentially mutated genes
  if(nrow(m2) == 0) m2 else{
    #creates matrix with 4 rows
    xx = with(m2, matrix(c(N.hightotal - N.high, N.high, N.lowtotal - N.low, N.low), 4, byrow = TRUE))

    #Performs Chi-sq tests for each gene, adjusts p-value, returns data frame
    m2[, p.value := apply(xx, 2, function(x) {
      oopts <- options(warn = -1)
      on.exit(oopts)
      (chisq.test(matrix(x, 2))$p.value)
    })]

    m2[, FDR := p.adjust(p.value, method = "BH")]
    setkey(m2, p.value)
    m2}
  }
}
