#' plotmut function
#'
#' Draws plot of mutations
#'
#' @param mae MultiAssayExperiment object. Must contain assays simplified Mutation (generated by `qreduceTCGA`)
#' or a binary matrix of mutated genes by sample, assay should contain "Mutation" in the name,
#' and Cohort (from `rnasubset` or `mutsubset`)
#' @param genemut data frame generated from diffmut function
#' @param gene character(1) gene of interest
#' @param n numeric(1) number of genes to plot default = 20
#'
#' @import data.table
#' @import ggplot2
#' @import reshape2
#'
#' @return oncoprint-like plot for top differentially mutated genes
#'
#' @examples
#' library(curatedTCGAData)
#' library(TCGAutils)
#' lusc <- curatedTCGAData("LUSC", c("Mutation", "RNASeq2GeneNorm", "GISTIC_ThresholdedByGene", "RPPAArray"), FALSE)
#' genome(lusc[[2]]) <- vapply(genome(lusc[[2]]), translateBuild, character(1L))
#' seqlevelsStyle(lusc[[2]]) <- "NCBI"
#' lusc2 <- qreduceTCGA(lusc)
#'
#' #remove normal samples and just focus on primaries
#' tnmae <- splitAssays(lusc2, c("01", "11"))
#'
#' lusc.t <- tnmae[, , grep("^01", names(tnmae))]
#' #split tumor and normal samples
#' lusc_tn <- splitAssays(lusc2, c("01", "11"))
#' #remake MultiAssayExperiment with only primary tumor samples
#' lusc_t <- lusc_tn[, , grep("^01", names(lusc_tn))]
#' lusc_egfr <- rnasubset(lusc_t, "EGFR", 10)
#' egfr_dm <- diffmut(lusc_egfr)
#' plotmut(lusc_egfr, egfr_dm, "EGFR", 20)
#'
#' @export
#'
plotmut <- function(mae, genemut, gene, n = 20) {
  #check only one mutation and one cohort assay in the multiassayexperiment object
  stopifnot(class(mae) == "MultiAssayExperiment", any(grepl("Mutation", names(mae))), any(grepl("Cohort", names(mae))))
  if(length(grep("Mutation", names(mae))) > 1)
    stop("Only one mutation assay is allowed")
  if(length(grep("Cohort", names(mae))) > 1)
    stop("Only one cohort comparison is allowed")

  # Find which assays contain mutation and expression level
  mut_assay <- grep("Mutation", names(mae))
  exp_assay <- grep("Cohort", names(mae))

  mae2 <- intersectColumns(mae[, , c(mut_assay, exp_assay)])

  annot <- dcast(as.data.frame(sampleMap(mae2)), primary ~ assay, value.var = "colname")
  lvl2 <- data.frame(Cohort = colnames(mae2[[2]]), Level = mae2[[2]][1,])
  lvl3 <- merge(annot, lvl2, by = "Cohort")

  lvl.high <- lvl3[lvl3$Level == "high", grep("Mutation", colnames(lvl3))]
  lvl.low <- lvl3[lvl3$Level == "low", grep("Mutation", colnames(lvl3))]

  mut <- assay(mae2[[1]])

  #Choose which genes to plot (num_genes)
  glist <- c(genemut$Gene[1:n], gene)
  glist <- glist[!duplicated(glist)]

  #limit mut matrix to genes chosen above
  mut.high <- mut[glist, lvl.high]
  mut.low <- mut[glist, lvl.low]

  #gets order of patients
  mhm <- melt(mut.high)
  gg1.high <- dcast(mhm, Var2 ~ Var1)
  setDT(gg1.high)
  gene.order.high <- names(sort(apply(gg1.high[,colnames(gg1.high)[-1], with = FALSE], 2, sum), decreasing=T))
  setkeyv(gg1.high, gene.order.high)
  mhm$level <- "high"

  mlm <- melt(mut.low)
  gg1.low <- dcast(mlm, Var2 ~ Var1)
  setDT(gg1.low)
  gene.order.low <- names(sort(apply(gg1.low[,colnames(gg1.low)[-1], with = FALSE], 2, sum), decreasing=T))
  setkeyv(gg1.low, gene.order.low)
  mlm$level <- "low"

  #combine mutation groups order patients according to sorting from individual dcasts
  mut.melt <- rbind(mhm, mlm)
  names(mut.melt) <- c("Gene", "barcode", "mutation", "level")

  mut.melt$barcode <- factor(mut.melt$barcode,
                             levels = c(as.character(rev(gg1.high$Var2)),
                                        as.character(rev(gg1.low$Var2))))
  mut.melt$Gene <- factor(mut.melt$Gene, levels = rev(gene.order.high))

  gg3 <- ggplot(mut.melt, aes(x = barcode, y = Gene, fill = as.factor(mutation))) +
    geom_tile(colour = "white") +
    labs(x = "Patient", y = "Gene") +
    theme(title = element_text(size = 16),
          axis.ticks = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_text(size = 12),
          axis.title.x = element_text(size = 14),
          axis.title.y = element_text(size = 14),
          legend.text = element_text(size = 12)) +
    scale_fill_brewer(type = "qual",
                      name = "Legend",
                      palette = 6,
                      labels = c("Wild-type", "Mutated")) +
    ggtitle(paste(gene)) +
    facet_grid(. ~ level, scales = "free", space = "free")
  gg3
  #ggplotly(gg3)
}
